### PR TITLE
Prototype of HTTP authentication challenges

### PR DESCRIPTION
**Pull request by markstory**
Use exceptions and make HTTP authentication failures prevent lower layers of middleware from being invoked until the HTTP authentication passes.

I've not updated digest authentication yet, and the test coverage is a bit sparse, but I'll update that once people are happy with the general approach.

Refs #47
Refs #48

### CODE REVIEW

File: src/Authenticator/HttpBasicAuthenticator.php [hunk 1]
```diff
@@ -72,32 +74,28 @@ public function getUser(ServerRequestInterface $request)
...
+     * Create a challenge exception for basic auth challenge.
      *
      * @param \Psr\Http\Message\ServerRequestInterface $request A request object.
-     * @param \Psr\Http\Message\ResponseInterface $response A response object.
-     * @return void
-     * @throws \Cake\Network\Exception\UnauthorizedException
+     * @throws \Authentication\Authenticator\ChallengeException
      */
-    public function unauthenticated(ServerRequestInterface $request, ResponseInterface $response)
+    public function authenticationChallenge(ServerRequestInterface $request)
```
Comment (burzum) [inline src/Authenticator/HttpBasicAuthenticator.php]: Is there really no case in which we might need to modify the response as well?

**Comment by burzum** [comment]
Looks OK to me.

**Comment by ADmad** [comment]
I have a concern about the name of exception and interface. Besides http basic and http digest auth for other stateless authenticators also a 401 response code should be returned albeit without any challenge headers. So maybe instead of word "Challenge" something more general can be used?

**Comment by markstory** [comment]
@ADmad I'm open to ideas. I couldn't come up with something better.

**Comment by ADmad** [comment]
How about `StatelessAuthenticatorInterface` (it is a big long) instead of `ChallengerInterface`?  All stateless authenticator would implement it. That would also remove the need for having `AuthenticateInterface::isStateless()` method. We could just check if an authenticator implements `StatelessAuthenticatorInterface`.

Also have a `UnauthenticatedException` in the plugin with stateless authenticators can throw. It could replace `ChallengeException` or `ChallengeException` can extend `UnauthenticatedException`.

**Comment by markstory** [comment]
`UnauthenticatedException` sounds good to me. If the interface uses `Stateless` in the name, should we also use that interface to decide whether or not an authenticator is stateless and drop the `isStateless()` method?

**Comment by ADmad** [comment]
> If the interface uses Stateless in the name, should we also use that interface to decide whether or not an authenticator is stateless and drop the isStateless() method?

We should IMO.

**Comment by markstory** [comment]
@ADmad updated.

**Comment by ADmad** [comment]
Looks good to me.

**Comment by lorenzo** [comment]
Does this mean that only one authentication method can be tried as opposed to running through all methods until one matches?

**Comment by markstory** [comment]
> Does this mean that only one authentication method can be tried as opposed to running through all methods until one matches?

Right now that is the case if you are using a stateless authenticator. One could configure the basic/digest auth as the 'last' authenticator though.

**Comment by markstory** [comment]
Not really sure what is going on with the PHP5.6 builds.

**Comment by burzum** [comment]
> Right now that is the case if you are using a stateless authenticator. One could configure the basic/digest auth as the 'last' authenticator though.

This should be added to the docs or people will run into this trap.

**Comment by ADmad** [comment]
@markstory Looks like some phpunit issue.

**Comment by markstory** [comment]
@burzum Another option is to try and work around it so the configuration order doesn't matter. Personally I think the documentation is the simpler of the two approaches for now.

**Comment by lorenzo** [comment]
Does any of you have had a case  requiring multiple stateless auth mechanisms? I remember in particular that I tend to avoid having basic auth in the stack because if the automatic headers sending, as I had no control over when it was ok to send them.

**Comment by ADmad** [comment]
> Does any of you have had a case requiring multiple stateless auth mechanisms?

I haven't.

**Comment by markstory** [comment]
@lorenzo I have not either. I've used either Basic or Digest Auth but never both.

**Comment by markstory** [comment]
Anything else I should fix up on this? I can update the documentation in the README if people think its necessary.

**Comment by burzum** [comment]
@markstory I think we should update the docs.

**Comment by markstory** [comment]
Docs updated 👍

**Pull request merged**